polartools/diffraction.py

```diff
@@ -0,0 +1,189 @@
+    """
+
+    models = {
+        "Gaussian": GaussianModel(),
+        "Lorentzian": LorentzianModel(),
+        "Voigt": VoigtModel(),
+        "PseudoVoigt": PseudoVoigtModel(),
+    }
+    try:
+        peak_mod = models[model]
+    except KeyError:
+        raise ValueError(
+            f"model = {model} is invalid. Only these values are accepted: "
+            f"{list(models.keys())}"
+        )
+    background = LinearModel()
+    mod = peak_mod + background
+    pars = background.make_params(intercept=ydata.min(), slope=0)
+    pars += peak_mod.guess(ydata, x=xdata)
+    pars["sigma"].set(min=0)
+    pars["amplitude"].set(min=0)
+
+    fit = mod.fit(ydata, pars, x=xdata)
+    print(f"Fitting with {model} model")
+    if output:
+        for key in fit.params:
+            print(
+                key, "=", fit.params[key].value, "+/-", fit.params[key].stderr
+            )
+        plt.plot(xdata, ydata)
+        plt.plot(xdata, fit.best_fit)
+        plt.show()
+    return fit
+
+
+def fit_series(
+    db,
+    scan_series,
+    model="Gaussian",
+    output=False,
+    var_series=None,
+    positioner="4C Theta",
+    detector="APD",
+    monitor=None,
+    **kwargs,
+):
+    """
+    Fit series of scan with reflection profile of choice and provide fit parameters as list.
+
+    Uses lmfit (https://lmfit.github.io/lmfit-py/).
+
+
+    Parameters
+    ----------
+    db : database
+        Databroker database. If None, it will attempt to read from csv or spec
+        files.
+    scan_series : int
+        start, stop, step, [start2, stop2, step2, ... ,startn, stopn, stepn]
+    model:
+        fit model: Gaussian, Lorentian, Voigt, PseidoVoigt
+    output:
+        Output fit parameters and plot data+fit for each scan.
+    var_series:
+        Varying variable for scan series, e.g. SampK (sample temperature), optional.
+    positioner : string
+        Name of the positioner, this needs to be the same as defined in
+        Bluesky.
+    detector : string
+        Detector to be read, needs to be the same name as in Bluesky.
+    monitor : string, optional
+        Name of the monitor detector. The returned scans will be
+        detectors/monitor.
+    kwargs :
+        Passed to `load_table` and 'fit_peak'.
+
+    Returns
+    -------
+    fit : lmfit ModelResult class
+        Contains the fit results. See:
+        https://lmfit.github.io/lmfit-py/model.html#the-modelresult-class
+    """
+
+    if len(scan_series) % 3:
+        raise ValueError(
+            f"expected 3*n={3*(len(scan_series)//3)} arguments, got {len(scan_series)}"
+        )
+
+    nbp = 0
+    for series in range(1, len(scan_series), 3):
+        nbp = (
+            nbp
+            + int(scan_series[series] - scan_series[series - 1])
+            / scan_series[series + 1]
+            + 1
+        )
+    fit_result = [np.zeros(7)]
+    for i in range(int(nbp) - 1):
+        fit_result.append(np.zeros(7))
+
```

Comment:
Suggestion:

The `load_spec` function was reloading the whole spec file for each scan, but this can be quite time consuming! So I changed it so that it now takes a `spec_file` which can be either a string (this will do the same as before), or it can be a `spec2nexus.spec.SpecDataFile` instance. So here, it would be faster to do (note that you need to import `from spec2nexus.spec import SpecDataFile`):
```
spec_file = kwargs.pop('spec_file', None)
if spec_file and isinstance(spec_file, str):
    spec_file = SpecDataFile(spec_file)
```
Then in line 151:
```
table = load_table(
                 scan, db, detector=detector, monitor=monitor, spec_file=spec_file, **kwargs
             )
```

Reply:
changed to load scan now